chall04/migferna.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+
+import sys
+import collections
+import os
+
+def order(bookshelfs, books):
+    bookshelfs = [int(x) for x in bookshelfs]
+    bookshelfs.sort(reverse=True)
+    books = sorted(books, key = lambda i: i['width'], reverse=True)
+    
+    i = 0
+    for book in books:
+        for i, bookshelf in enumerate(bookshelfs):
+            if book['width'] <= bookshelf:
+                bookshelf -= book['width']
+                bookshelfs[i] = bookshelf
+                break
+            elif len(bookshelfs) == i + 1:
+                return
+            else:
+                bookshelf + 1
+    return i if len(bookshelfs) == 0 else i+1
```

Comment:
I had trouble understanding what you did here. There is a simpler and more readable way to reproduce this behavior.  
I understand you want to iterate on books and shelves, however the `for ... in ...` loop is not what you need here as you can't iterate according to a condition.

You might want to create an iterator for your shelves before entering the books loop, it would look like this
```python3
shelves_iter = iter(bookshelfs)
current_shelf = next(shelves_iter) # This is to initiate the variable and get the first shelf
for book in books:
    if shelf >= book["width"]:
        # Handle a book that fits
        continue # You stored the book, you may now continue to the next book
    try:
        shelf = next(shelf)
        # Validate your new shelf
    except StopIteration: # next(iterator) raises this exception when it can't iterate anymore
        break
```

Later I advise you to look at the `for ... else ...` syntax to handle your return values better